pkg/sqlite/deprecate.go

```diff
@@ -45,3 +48,82 @@ func (d *BundleDeprecator) Deprecate() error {
 
 	return utilerrors.NewAggregate(errs)
 }
+
+// MaybeRemovePackages queries the DB to establish if any provided bundles are the head of the default channel of a package.
+// If so, the list of bundles must also contain the head of all other channels in the package, otherwise an error is produced.
+// If the heads of all channels are being deprecated (including the default channel), the package is removed entirely from the index.
+// MaybeRemovePackages deletes all bundles from the associated package from the bundles array, so that the subsequent
+// Deprecate() call can proceed with deprecating other potential bundles from other packages.
+func (d *BundleDeprecator) MaybeRemovePackages() error {
+	log := logrus.WithField("bundles", d.bundles)
+	log.Info("allow-package-removal enabled: checking default channel heads for package removal")
+
+	var errs []error
+	var removedBundlePaths []string
+	var remainingBundlePaths []string
+
+	// Iterate over bundles list - see if any bundle is the head of a default channel in a package
+	var packages []string
+	for _, bundle := range d.bundles {
+		found, err := d.querier.PackageFromDefaultChannelHeadBundle(context.TODO(), bundle)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error checking if bundle is default channel head %s: %s", bundle, err))
+		}
+		if found != "" {
+			packages = append(packages, found)
+		}
+	}
+
+	if len(packages) == 0 {
+		return nil
+	}
+
+	// If so, ensure list contains head of all other channels in that package
+	// If not, return error
+	for _, pkg := range packages {
+		channels, err := d.querier.ListChannels(context.TODO(), pkg)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error listing channels for package %s: %s", pkg, err))
+		}
+		for _, channel := range channels {
+			found, err := d.querier.BundlePathForChannelHead(context.TODO(), pkg, channel)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("error listing channel head for package %s: %s", pkg, err))
+			}
+			if !contains(found, d.bundles) {
+				return fmt.Errorf("cannot deprecate default channel head from package without removing all other channel heads in package %s: must deprecate %s, head of channel %s", pkg, found, channel)
+			}
+			removedBundlePaths = append(removedBundlePaths, found)
+		}
+	}
+
+	// Remove associated package from index
+	for _, pkg := range packages {
+		err := d.store.RemovePackage(pkg)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error removing package %s: %s", pkg, err))
+		}
+	}
+
+	// Remove bundles from the removed package from the deprecation request
+	// This enables other bundles to be deprecated via the expected flow
+	// Build a new array with just the outstanding bundles
+	for _, bundlePath := range d.bundles {
+		if contains(bundlePath, removedBundlePaths) {
+			continue
+		}
+		remainingBundlePaths = append(remainingBundlePaths, bundlePath)
+	}
+	d.bundles = remainingBundlePaths
```

Comment:
I'm not sure I understand what this does. This is the only function using the bundleDeprecator right?

Reply:
No, there is also the existing `Deprecate()` logic which I did not change. This line ensures that the `Deprecate()` call will run on a subset of packages after the `MaybeRemoveDeprecatedPackages()` call returns. 

For example, say that there are three bundles to deprecatetruncate A B C. A and B are in one package, A is the head of the default channel, B is the head of the only other channel. This package would be removed entirely from the index, but C would also need to be deprecated via the existing logic (find tail bundles, remove those, add the deprecation property, etc). So what this logic does is take out bundles that were associated with packages that were then removed (A and B), leaving only those that should still be deprecated via the expected flow (C)

Reply:
gotcha thanks!